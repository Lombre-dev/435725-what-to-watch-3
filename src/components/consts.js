export const GENRES = [
  `Comedy`,
  `Crime`,
  `Documentary`,
  `Drama`,
  `Horror`,
  `Kids & Family`,
  `Romance`,
  `Sci-Fi`,
  `Thriller`,
];

export const RATING_SCORE_LIMIT = 10;
export const RATING_LEVELS = [
  `Bad`,
  `Normal`,
  `Good`,
  `Very good`,
  `Awesome`,
];
