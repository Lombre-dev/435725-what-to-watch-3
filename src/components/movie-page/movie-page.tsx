import * as React from 'react';
import {connect} from 'react-redux';
import {Redirect, withRouter} from 'react-router-dom';
import {setDetailedMovieRedirectTo} from '../../redux/movie/actions';
import {Operations} from '../../redux/movie/operations';
import {getDetailedMovie, getDetailedMovieRedirectTo, getDetailedMovieRelatedMovies, getDetailedMovieStatus} from '../../redux/movie/selectors';
import BigMovieCard from '../big-movie-card/big-movie-card';
import Footer from '../footer/footer';
import LoadingDataBlock from '../loading-data-block/loading-data-block';
import Logo from '../logo/logo';
import MoreLikeThis from '../more-like-this/more-like-this';
import MovieInfo from '../movie-info';
import UserBlock from '../user-block/user-block';

type TMoviePageProps = {
  match: TMatchParamsWithId,
  status?: LoadingDataStatus,
  redirectTo?: string,
  movie?: TMovie,
  relatedMovies?: TMovie[],

  updateMovie?: Function,
  resetRedirect?: Function,
};

class MoviePage extends React.PureComponent<TMoviePageProps> {

  public componentDidMount() {

    const {updateMovie, match} = this.props;

    updateMovie(match.params.id);
  }

  public componentDidUpdate(prevProps) {

    const {match: {params: {id: currentId}}, updateMovie} = this.props;
    const {match: {params: {id: prevId}}} = prevProps;

    if (currentId !== prevId) {
      updateMovie(currentId);
    }
  }

  public componentWillUnmount() {

    const {resetRedirect} = this.props;

    resetRedirect();
  }

  public render() {

    const {redirectTo, status, movie, relatedMovies} = this.props;

    if (redirectTo) {
      return <Redirect to={redirectTo} />;
    }

    if (status !== LoadingDataStatus.READY && !movie) {
      return <LoadingDataBlock status={status} />;
    }

    return (
      <>
        <section className="movie-card movie-card--full">
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={movie.backgroundImage} alt={movie.title} />
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <header className="page-header movie-card__head">
              <Logo />
              <UserBlock />
            </header>

            <div className="movie-card__wrap">
              <BigMovieCard
                movie={movie}
              />
            </div>
          </div>
          <MovieInfo
            movie={movie}
          />
        </section>

        <div className="page-content">
          <MoreLikeThis
            movies={relatedMovies}
          />
          <Footer />
        </div>
      </>
    );
  }
}

function mapStateToProps(state: Object) {
  return {
    status: getDetailedMovieStatus(state),
    redirectTo: getDetailedMovieRedirectTo(state),
    movie: getDetailedMovie(state),
    relatedMovies: getDetailedMovieRelatedMovies(state),
  };
}

function mapDispatchToProps(dispatch: Object) {
  return {
    updateMovie: (movieId) => {
      dispatch(Operations.init(movieId));
    },
    resetRedirect: () => {
      dispatch(setDetailedMovieRedirectTo(undefined));
    },
  };
}

export {MoviePage};
export default connect(mapStateToProps, mapDispatchToProps)(withRouter(MoviePage));
