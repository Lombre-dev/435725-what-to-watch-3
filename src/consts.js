export const ALL_GENRE = `All genres`;
const COMEDY_GENRE = `Comedy`;
const CRIME_GENRE = `Crime`;
const DOCUMENTARY_GENRE = `Documentary`;
const DRAMA_GENRE = `Drama`;
const HORROR_GENRE = `Horror`;
const KIDS_AND_FAMILY_GENRE = `Kids & Family`;
const ROMANCE_GENRE = `Romance`;
const SCI_FI_GENRE = `Sci-Fi`;
const THRILLER_GENRE = `Thriller`;

export const AppPages = {
  MAIN: `/`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  MOVIES: `/films`,
  PLAYER: `/player`,
};

export const GENRES = [
  COMEDY_GENRE,
  CRIME_GENRE,
  DOCUMENTARY_GENRE,
  DRAMA_GENRE,
  HORROR_GENRE,
  KIDS_AND_FAMILY_GENRE,
  ROMANCE_GENRE,
  SCI_FI_GENRE,
  THRILLER_GENRE,
];

export const GenreLabels = {
  [COMEDY_GENRE]: `Comedies`,
  [DRAMA_GENRE]: `Dramas`,
  [THRILLER_GENRE]: `Thrillers`,
};

export const MORE_LIKE_THIS_MOVIES_LIMIT = 4;
export const CATALOG_MOVIES_PER_PAGE_LIMIT = 8;

export const SMALL_MOVIE_CARD_PREVIEW_DELAY = 1000;

export const MOVIE_INFO_TABS = [
  `Overview`,
  `Details`,
  `Reviews`,
];

export const PlayerState = {
  INITED: `inited`,
  LOADING: `loading`,
  PLAYING: `playing`,
  PAUSED: `paused`,
  ENDED: `ended`,
};

export const SERVER_URL = `https://htmlacademy-react-3.appspot.com/wtw`;
export const SERVER_RESPONSE_TIMEOUT = 5000;
export const SERVER_USE_COOKIES = true;
export const ServerErrors = {
  NOT_FOUND: 404,
  UNAUTHORIZED: 401,
};
