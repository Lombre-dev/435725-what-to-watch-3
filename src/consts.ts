export const ALL_GENRE = `All genres`;
const COMEDY_GENRE = `Comedy`;
const CRIME_GENRE = `Crime`;
const DOCUMENTARY_GENRE = `Documentary`;
const DRAMA_GENRE = `Drama`;
const HORROR_GENRE = `Horror`;
const KIDS_AND_FAMILY_GENRE = `Kids & Family`;
const ROMANCE_GENRE = `Romance`;
const SCI_FI_GENRE = `Sci-Fi`;
const THRILLER_GENRE = `Thriller`;

export enum AppPages {
  MAIN = `/`,
  LOGIN = `/login`,
  MY_LIST = `/mylist`,
  MOVIES = `/films`,
  PLAYER = `/player`,
}

export const GENRES = [
  COMEDY_GENRE,
  CRIME_GENRE,
  DOCUMENTARY_GENRE,
  DRAMA_GENRE,
  HORROR_GENRE,
  KIDS_AND_FAMILY_GENRE,
  ROMANCE_GENRE,
  SCI_FI_GENRE,
  THRILLER_GENRE,
];
export const GENRES_CATALOG_LIMIT = 9;

export const GenreLabels = {
  [COMEDY_GENRE]: `Comedies`,
  [DRAMA_GENRE]: `Dramas`,
  [THRILLER_GENRE]: `Thrillers`,
};

export const MORE_LIKE_THIS_MOVIES_LIMIT = 4;
export const CATALOG_MOVIES_PER_PAGE_LIMIT = 8;

export const SMALL_MOVIE_CARD_PREVIEW_DELAY = 1000;

export const MOVIE_INFO_TABS = [
  `Overview`,
  `Details`,
  `Reviews`,
];

export enum PlayerState {
  INITED = `inited`,
  LOADING = `loading`,
  PLAYING = `playing`,
  PAUSED = `paused`,
  ENDED = `ended`,
}

export const ServerConfig = {
  URL: `https://htmlacademy-react-3.appspot.com/wtw`,
  RESPONSE_TIMEOUT: 5000,
  USE_COOKIES: true,
};
export enum ServerErrors {
  NOT_FOUND = 404,
  BAD_REQUEST = 400,
  UNAUTHORIZED = 401,
}

export enum AuthorizationErrorCode {
  INCORRECT_LOGIN = 1,
  AUTHORIZATION_FAIL = 2,
}

export const AuthorizationErrorMessage = {
  [AuthorizationErrorCode.INCORRECT_LOGIN]: `Please enter a valid email address`,
  [AuthorizationErrorCode.AUTHORIZATION_FAIL]: `We canâ€™t recognize this email and password combination. Please try again.`,
};

export enum AuthorizationStatus {
  NO_AUTH = `unauthorized`,
  AUTH = `authorized`,
}

export const REVIEW_RATING_STARS_LIMIT = 5;
export const REVIEW_COMMENT_MIN_LENGTH = 50;
export const REVIEW_COMMENT_MAX_LENGTH = 400;

export enum LoadingDataStatus {
  LOADING = `loading`,
  READY = `ready`,
  ERROR = `error`,
}
