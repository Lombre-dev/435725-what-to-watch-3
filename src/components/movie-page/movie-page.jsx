import PropTypes from 'prop-types';
import React from 'react';
import {connect} from 'react-redux';
import {Redirect, withRouter} from 'react-router-dom';
import {LoadingDataStatus} from '../../consts';
import {setDetailedMovieRedirectTo} from '../../redux/movie/actions';
import {Operations} from '../../redux/movie/operations';
import {getDetailedMovie, getDetailedMovieRedirectTo, getDetailedMovieRelatedMovies, getDetailedMovieStatus} from '../../redux/movie/selectors';
import BigMovieCard from '../big-movie-card/big-movie-card';
import Footer from '../footer/footer';
import LoadingDataBlock from '../loading-data-block/loading-data-block';
import Logo from '../logo/logo';
import MoreLikeThis from '../more-like-this/more-like-this';
import MovieInfo from '../movie-info';
import {Movie} from '../types';
import UserBlock from '../user-block/user-block';

class MoviePage extends React.PureComponent {

  componentDidMount() {

    const {onMount, match: {params: {id}}} = this.props;

    onMount(id);
  }

  componentDidUpdate(prevProps) {

    const {match: {params: {id: currentId}}, onMount} = this.props;
    const {match: {params: {id: prevId}}} = prevProps;

    if (currentId !== prevId) {
      onMount(currentId);
    }
  }

  componentWillUnmount() {

    const {onUnmount} = this.props;

    onUnmount();
  }

  render() {

    const {redirectTo, status, movie, relatedMovies} = this.props;

    if (redirectTo) {
      return <Redirect to={redirectTo} />;
    }

    if (status !== LoadingDataStatus.READY && !movie) {
      return <LoadingDataBlock status={status} />;
    }

    return (
      <>
        <section className="movie-card movie-card--full">
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={movie.backgroundImage} alt={movie.title} />
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <header className="page-header movie-card__head">
              <Logo />
              <UserBlock />
            </header>

            <div className="movie-card__wrap">
              <BigMovieCard
                movie={movie}
              />
            </div>
          </div>
          <MovieInfo
            movie={movie}
          />
        </section>

        <div className="page-content">
          <MoreLikeThis
            movies={relatedMovies}
          />
          <Footer />
        </div>
      </>
    );
  }
}

MoviePage.propTypes = {
  match: PropTypes.object.isRequired,
  status: PropTypes.oneOf(Object.values(LoadingDataStatus)),
  redirectTo: PropTypes.string,
  movie: Movie,
  relatedMovies: PropTypes.arrayOf(Movie),

  onMount: PropTypes.func,
  onUnmount: PropTypes.func,
};

function mapStateToProps(state) {
  return {
    status: getDetailedMovieStatus(state),
    redirectTo: getDetailedMovieRedirectTo(state),
    movie: getDetailedMovie(state),
    relatedMovies: getDetailedMovieRelatedMovies(state),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onMount: (movieId) => {
      dispatch(Operations.init(movieId));
    },
    onUnmount: () => {
      dispatch(setDetailedMovieRedirectTo(undefined));
    },
  };
}

export {MoviePage};
export default connect(mapStateToProps, mapDispatchToProps)(withRouter(MoviePage));
