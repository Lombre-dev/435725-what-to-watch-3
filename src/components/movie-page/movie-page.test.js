import React from 'react';
import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {AuthorizationStatus, LoadingDataStatus} from '../../consts';
import {MoviePage} from './movie-page';

const MATCH = {params: {id: 0}};
const HANDLE_EVENT = () => {};
const MOVIE = {
  id: 0,
  title: `The Grand Budapest Hotel`,
  genres: [`Drama`],
  year: 2014,
  poster: `/img/the-grand-budapest-hotel-poster.jpg`,
  frames: [`/img/the-grand-budapest-hotel-poster.jpg`],
  ratingScore: 8.9,
  ratingReviewsCount: 240,
  preview: `/samples/sintel_trailer-480p.mp4`,
  src: `/samples/sintel_trailer-480p.mp4`,
  description: `Description`,
  story: `Story`,
  director: `Wes Andreson`,
  duration: 99,
  actors: [
    `Bill Murray`,
    `Edward Norton`,
    `Jude Law`,
    `Willem Dafoe`,
    `Some Actor 1`,
    `Some Actor 2`,
  ],
};
const MOVIES_LIKE_CURRENT = [];
const LOADING_DATA_STATUS = LoadingDataStatus.READY;
const mockStore = configureStore([thunk]);

describe(`<MoviePage />`, () => {

  it(`render should be match markup`, () => {

    const store = mockStore({
      user: {
        authState: AuthorizationStatus.AUTH,
      },
    });

    const result = renderer
      .create(<Provider store={store}>
        <BrowserRouter>
          <MoviePage
            match={MATCH}
            statis={LOADING_DATA_STATUS}
            movie={MOVIE}
            moviesLikeCurrent={MOVIES_LIKE_CURRENT}
            init={HANDLE_EVENT}
          />
        </BrowserRouter>
      </Provider>)
      .toJSON();

    expect(result).toMatchSnapshot();
  });
});
