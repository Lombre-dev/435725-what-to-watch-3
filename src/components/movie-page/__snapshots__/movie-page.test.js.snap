// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoviePage /> render should be match markup 1`] = `<p />`;
